meteor/map_utils.py

```diff
@@ -0,0 +1,216 @@
+
+    result = opt.least_squares(
+        aniso_scale_func,
+        initial_params,
+        args=(reference_data, scale_data, miller_indices_scale),
+    )
+
+    # Apply the scaling to dataset_to_scale
+    h, k, l = miller_indices_scale[:, 0], miller_indices_scale[:, 1], miller_indices_scale[:, 2]  # noqa: E741
+    h_sq, k_sq, l_sq = np.square(h), np.square(k), np.square(l)
+    hk_prod, hl_prod, kl_prod = h * k, h * l, k * l
+
+    t = -(
+        h_sq * result.x[1]
+        + k_sq * result.x[2]
+        + l_sq * result.x[3]
+        + 2 * hk_prod * result.x[4]
+        + 2 * hl_prod * result.x[5]
+        + 2 * kl_prod * result.x[6]
+    )
+
+    scale_factor = result.x[0] * np.exp(t)
+    scaled_data = scale_factor * scale_data
+
+    if uncertainties is not None:
+        scaled_uncertainties = scale_factor * uncertainties_data
+    else:
+        scaled_uncertainties = None
+
+    if inplace:
+        dataset_to_scale[:] = scaled_data
+        if uncertainties is not None:
+            uncertainties[:] = scaled_uncertainties
+        return None
+    else:
+        scaled_dataset = dataset_to_scale.copy()
+        scaled_dataset[:] = scaled_data
+        if uncertainties is not None:
+            scaled_uncertainties_series = uncertainties.copy()
+            scaled_uncertainties_series[:] = scaled_uncertainties
+            return scaled_dataset, scaled_uncertainties_series
+        return scaled_dataset
+
+def compute_amplitude_fofo_difference(
+    data1: rs.DataSeries,
+    data2: rs.DataSeries,
+    data3: rs.DataSeries,
+    uncertainties1: Optional[rs.DataSeries] = None,
+    uncertainties2: Optional[rs.DataSeries] = None,
+) -> Union[rs.DataSeries, tuple[rs.DataSeries, rs.DataSeries]]:
+    """
+    First, scale data1 and data2 to the common scale defined by data3,
+    then compute the difference (data2 - data1). Optionally, propagate uncertainties.
+
+    Parameters:
+    data1 (rs.DataSeries): First dataset to be used in difference calculation (e.g., F_off).
+    data2 (rs.DataSeries): Second dataset to be used in difference calculation (e.g., F_on).
+    data3 (rs.DataSeries): Reference dataset used for scaling data1 and data2 (e.g., F_calc).
+    uncertainties1 (Optional[rs.DataSeries]): Uncertainties corresponding to data1 (e.g. SIGF_off).
+    uncertainties2 (Optional[rs.DataSeries]): Uncertainties corresponding to data2 (e.g. SIGF_on).
+
+    Returns:
+    rs.DataSeries or tuple: The difference (data2 - data1) after scaling to the reference scale.
+    If uncertainties are provided, returns a tuple (difference, propagated_uncertainties).
+    """
+
+    # Ensure that both or neither uncertainties are provided
+    if (uncertainties1 is not None) != (uncertainties2 is not None):
+        raise ValueError("Either provide uncertainties for both data1 and data2, or provide none.")
+
+    if uncertainties1 is not None and uncertainties2 is not None:
```

Comment:
given all the hoops you have to jump through to handle the with and without uncertainties thing, I would just make two different functions -- one taking the uncertainties and one not